### ISSUE TITLE
Unable to upload the image due to "open /scratch/tmpimage: permission denied"

### ISSUE THREAD

**Issue by manaspal2** [kubevirt/containerized-data-importer#2739]
**What happened**:
I have a 2 bare metal server K8S cluster. Both the nodes are running on CBL-Meriner 2.0.
My plan is to use Kubevirt to launch my VM. I have created PV and DV. After that, when i see that PVC status is Bound, I am trying to upload my image (qcow2). But face the permission problem.

root [ /home/admin ]# virtctl image-upload dv vsrx-boot --size=25Gi --image-path=/home/admin/junos-vsrx3-x86-64-23.4I-20230531_dev_common.0.1657.qcow2
 --uploadproxy-url https://10.244.1.84:8443  --insecure --storage-class hostpath-provisioner -v9
Using existing PVC default/vsrx-boot
Uploading data to https://10.244.1.84:8443

 1.06 MiB / 831.75 MiB [>-----------------------------------------------------------------------------------------------------------------]   0.13% 0s

unexpected return value 500, Saving stream failed: Unable to transfer source data to scratch space: could not open file "/scratch/tmpimage": open /scratch/tmpimage: permission denied

**What you expected to happen**:
I expect here something like below;
Using existing PVC default/vsrx-boot
Uploading data to https://10.244.1.10:8443

 746.81 MiB / 746.81 MiB [=================================================================================================] 100.00% 6s

Uploading data completed successfully, waiting for processing to complete, you can hit ctrl-c without interrupting the progress
Processing completed successfully
Uploading vsrx3_21.4R2_kubevirt.qcow2 completed successfully

**How to reproduce it (as minimally and precisely as possible)**:
Install K8S v1.26.
Install Kubevirt v0.59.0
Install CDI version v1.56.0
Create PV and DV.
After that try to upload the image.

**Additional context**:
root@pf-cloud$ kubectl get pv
NAME     CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS   CLAIM                       STORAGECLASS           REASON   AGE
pv0001   25Gi       RWO            Retain           Bound    default/vsrx-boot-scratch   hostpath-provisioner            19h
pv0002   25Gi       RWO            Retain           Bound    default/vsrx-boot           hostpath-provisioner            19h

root@pf-cloud$ kubectl get pvc
NAME                STATUS   VOLUME   CAPACITY   ACCESS MODES   STORAGECLASS           AGE
vsrx-boot           Bound    pv0002   25Gi       RWO            hostpath-provisioner   18h
vsrx-boot-scratch   Bound    pv0001   25Gi       RWO            hostpath-provisioner   18h

root@pf-cloud$ kubectl get pods -o wide
NAME                   READY   STATUS    RESTARTS   AGE   IP            NODE                           NOMINATED NODE   READINESS GATES
cdi-upload-vsrx-boot   1/1     Running   0          18h   10.244.1.88   la-lakers.englab.juniper.net   <none>           <none>
root@pf-cloud$
root@pf-cloud$ kubectl describe pods/cdi-upload-vsrx-boot
Name:             cdi-upload-vsrx-boot
Namespace:        default
Priority:         0
Service Account:  default
Node:             la-lakers.englab.juniper.net/40.40.40.2
Start Time:       Thu, 01 Jun 2023 23:11:08 +0000
Labels:           app=containerized-data-importer
                  app.kubernetes.io/component=storage
                  app.kubernetes.io/managed-by=cdi-controller
                  cdi.kubevirt.io=cdi-upload-server
                  cdi.kubevirt.io/uploadTarget=996e5882-c401-4c92-b1e7-46642f0737aa
                  service=cdi-upload-vsrx-boot
Annotations:      cdi.kubevirt.io/storage.createdByUploadController: yes
                  sidecar.istio.io/inject: false
Status:           Running
IP:               10.244.1.88
IPs:
  IP:           10.244.1.88
Controlled By:  PersistentVolumeClaim/vsrx-boot
Containers:
  cdi-upload-server:
    Container ID:  containerd://43d1f09125a1f9e8fe6455fcfc3789a9e461f3286e94b4805606584d380a9c20
    Image:         quay.io/kubevirt/cdi-uploadserver:v1.56.0
    Image ID:      quay.io/kubevirt/cdi-uploadserver@sha256:b45d036b6eff2528a80f32ff5aad3e5c10b5d88821ddf88ee11f8a0e9b380669
    Port:          <none>
    Host Port:     <none>
    Args:
      -v=1
    State:          Running
      Started:      Thu, 01 Jun 2023 23:11:13 +0000
    Ready:          True
    Restart Count:  0
    Limits:
      cpu:     750m
      memory:  600M
    Requests:
      cpu:      100m
      memory:   60M
    Readiness:  http-get http://:8080/healthz delay=2s timeout=1s period=5s #success=1 #failure=3
    Environment:
      TLS_KEY:              <set to the key 'tls.key' in secret 'cdi-upload-vsrx-boot'>  Optional: false
      TLS_CERT:             <set to the key 'tls.crt' in secret 'cdi-upload-vsrx-boot'>  Optional: false
      CLIENT_CERT:          -----BEGIN CERTIFICATE-----
                            MIIDVTCCAj2gAwIBAgIIak2h9QJIlz0wDQYJKoZIhvcNAQELBQAwODE2MDQGA1UE
                            AwwtY2RpX2NkaS11cGxvYWRzZXJ2ZXItY2xpZW50LXNpZ25lckAxNjg0Nzg3MzQ2
                            MB4XDTIzMDUyMjIwMjkwNVoXDTMzMDUxOTIwMjkwNlowODE2MDQGA1UEAwwtY2Rp
                            X2NkaS11cGxvYWRzZXJ2ZXItY2xpZW50LXNpZ25lckAxNjg0Nzg3MzQ2MIIBIjAN
                            BgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAyZ7CCUWian8It+5CvLSpE8L7x/lZ
                            0Gk38V3Hg5jFy5FpcDML0PRwFWBxmvgKECZsQ2CvkQmfWtQxCH8wkkBJ8XKun2BV
                            9kuUEpZ9KXzKeqonfy6GO1T7zGrWt0r45LwQVvUXl/jbQS79pMWAtiZMRYz4nu7R
                            dk2uP5JoRwHQ78bYG3AAcJ3SiWV5pt1X5EMtIalzZvJV9ZCNL3A6ozOUQZox+Z7Z
                            moctY/n0uwAuc16+BpoLPX/y1Y87aSYLhvb0LAi1/Te0DIUxXlP71NLnDZYVvKUp
                            jcJemiQdnCZj28FpPIhTF5P9IrD/7LY1B9iYp0vr7hV+WYu5a9fv0FYq7QIDAQAB
                            o2MwYTAOBgNVHQ8BAf8EBAMCAqQwDwYDVR0TAQH/BAUwAwEB/zAdBgNVHQ4EFgQU
                            0POYR+TZJ5WiyMOLIxzl5s1qDw4wHwYDVR0jBBgwFoAU0POYR+TZJ5WiyMOLIxzl
                            5s1qDw4wDQYJKoZIhvcNAQELBQADggEBAHG4oEUon1AzpaFfeyc5QtAQw1PdGngm
                            y9bxWjUZp1145QIWVOcCaDrvdGk9p89Aopx2+ueovVZpvE2RLXypPtQDa7y5rU0f
                            V90STXJfnmJ3EyLWUIBxZOlbg0ynXDbhYfnzFgqqSU8YCC2QH+bXMe1D+NbV0NS2
                            f8oKbuaTxKBW5ZKBKiFqhUTmYWEu32shA40BHWoc7sz0fHkHBQR59gp4qdkqx7I3
                            6hJQze8RcMZSbyLoaUFrWXSnqogQ6XEhWkBTyIyowqHVqykhMJPpTgIx+cOhRL2P
                            T0MEHRUXZr+KtxCkyaVuK4rBzVcNQOyuiHzT388aQgU4qBBZCpQpL4I=
                            -----END CERTIFICATE-----

      FILESYSTEM_OVERHEAD:  0.055
      UPLOAD_IMAGE_SIZE:    25Gi
      CLIENT_NAME:          client.upload-server.cdi.kubevirt.io
      PREALLOCATION:        false
      TLS_CIPHERS:          TLS_AES_128_GCM_SHA256,TLS_AES_256_GCM_SHA384,TLS_CHACHA20_POLY1305_SHA256,ECDHE-ECDSA-AES128-GCM-SHA256,ECDHE-RSA-AES128-GCM-SHA256,ECDHE-ECDSA-AES256-GCM-SHA384,ECDHE-RSA-AES256-GCM-SHA384,ECDHE-ECDSA-CHACHA20-POLY1305,ECDHE-RSA-CHACHA20-POLY1305,DHE-RSA-AES128-GCM-SHA256,DHE-RSA-AES256-GCM-SHA384
      TLS_MIN_VERSION:      VersionTLS12
    Mounts:
      /data from cdi-data-vol (rw)
      /scratch from cdi-scratch-vol (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-5x4sr (ro)
Conditions:
  Type              Status
  Initialized       True
  Ready             True
  ContainersReady   True
  PodScheduled      True
Volumes:
  cdi-data-vol:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  vsrx-boot
    ReadOnly:   false
  cdi-scratch-vol:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  vsrx-boot-scratch
    ReadOnly:   false
  kube-api-access-5x4sr:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   Burstable
Node-Selectors:              kubernetes.io/os=linux
Tolerations:                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                             node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:                      <none>
root@pf-cloud$

Upload server log

I0602 16:49:49.929345       1 uploadserver.go:324] Content type header is ""
I0602 16:49:49.929390       1 data-processor.go:379] Calculating available size
I0602 16:49:49.929414       1 data-processor.go:391] Checking out file system volume size.
I0602 16:49:49.929434       1 data-processor.go:399] Request image size not empty.
I0602 16:49:49.929449       1 data-processor.go:404] Target size 25Gi.
I0602 16:49:49.929487       1 data-processor.go:282] New phase: TransferScratch
E0602 16:49:49.929555       1 data-processor.go:278] open /scratch/tmpimage: permission denied
could not open file "/scratch/tmpimage"
kubevirt.io/containerized-data-importer/pkg/util.OpenFileOrBlockDevice
	pkg/util/util.go:179
kubevirt.io/containerized-data-importer/pkg/util.StreamDataToFile
	pkg/util/util.go:186
kubevirt.io/containerized-data-importer/pkg/importer.(*AsyncUploadDataSource).Transfer
	pkg/importer/upload-datasource.go:146
kubevirt.io/containerized-data-importer/pkg/importer.(*DataProcessor).initDefaultPhases.func2
	pkg/importer/data-processor.go:208
kubevirt.io/containerized-data-importer/pkg/importer.(*DataProcessor).ProcessDataWithPause
	pkg/importer/data-processor.go:275
kubevirt.io/containerized-data-importer/pkg/uploadserver.newAsyncUploadStreamProcessor
	pkg/uploadserver/uploadserver.go:432
kubevirt.io/containerized-data-importer/pkg/uploadserver.(*uploadServerApp).uploadHandlerAsync.func1
	pkg/uploadserver/uploadserver.go:331
net/http.HandlerFunc.ServeHTTP
	GOROOT/src/net/http/server.go:2084
net/http.(*ServeMux).ServeHTTP
	GOROOT/src/net/http/server.go:2462
kubevirt.io/containerized-data-importer/pkg/uploadserver.(*uploadServerApp).ServeHTTP
	pkg/uploadserver/uploadserver.go:260
net/http.serverHandler.ServeHTTP
	GOROOT/src/net/http/server.go:2916
net/http.(*conn).serve
	GOROOT/src/net/http/server.go:1966
runtime.goexit
	GOROOT/src/runtime/asm_amd64.s:1571
Unable to transfer source data to scratch space
kubevirt.io/containerized-data-importer/pkg/importer.(*DataProcessor).initDefaultPhases.func2
	pkg/importer/data-processor.go:213
kubevirt.io/containerized-data-importer/pkg/importer.(*DataProcessor).ProcessDataWithPause
	pkg/importer/data-processor.go:275
kubevirt.io/containerized-data-importer/pkg/uploadserver.newAsyncUploadStreamProcessor
	pkg/uploadserver/uploadserver.go:432
kubevirt.io/containerized-data-importer/pkg/uploadserver.(*uploadServerApp).uploadHandlerAsync.func1
	pkg/uploadserver/uploadserver.go:331
net/http.HandlerFunc.ServeHTTP
	GOROOT/src/net/http/server.go:2084
net/http.(*ServeMux).ServeHTTP
	GOROOT/src/net/http/server.go:2462
kubevirt.io/containerized-data-importer/pkg/uploadserver.(*uploadServerApp).ServeHTTP
	pkg/uploadserver/uploadserver.go:260
net/http.serverHandler.ServeHTTP
	GOROOT/src/net/http/server.go:2916
net/http.(*conn).serve
	GOROOT/src/net/http/server.go:1966
runtime.goexit
	GOROOT/src/runtime/asm_amd64.s:1571
E0602 16:49:49.929743       1 uploadserver.go:336] Saving stream failed: Unable to transfer source data to scratch space: could not open file "/scratch/tmpimage": open /scratch/tmpimage: permission denied

cdi-uploadproxy log

I0602 16:49:49.858873       1 uploadproxy.go:204] Received valid token: pvc: vsrx-boot, namespace: default
I0602 16:49:49.899595       1 uploadproxy.go:204] Received valid token: pvc: vsrx-boot, namespace: default

**Environment**:
- CDI version (use `kubectl get deployments cdi-deployment -o yaml`): v1.56.0
- Kubernetes version (use `kubectl version`): 1.26
- DV specification: N/A
- Cloud provider or hardware configuration: N/A
- OS (e.g. from /etc/os-release): CBL-Mariner 2.0
- Kernel (e.g. `uname -a`): 5.15.102.1-3.cm2 #1 SMP Fri Apr 7 08:23:56 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
- Install tools: kubeadm
- Others: N/A

**Comment by awels** [comment]
```E0602 16:49:49.929743 1 uploadserver.go:336] Saving stream failed: Unable to transfer source data to scratch space: could not open file "/scratch/tmpimage": open /scratch/tmpimage: permission denied``` is the problem. What kind of storage do you have?

**Comment by manaspal2** [comment]
> `E0602 16:49:49.929743 1 uploadserver.go:336] Saving stream failed: Unable to transfer source data to scratch space: could not open file "/scratch/tmpimage": open /scratch/tmpimage: permission denied` is the problem. What kind of storage do you have?

Sorry, did not under your question. But probably if the following outputs help to answer your question.

root@pf-cloud$ kubectl describe pvc/vsrx-boot
Name:          vsrx-boot
Namespace:     default
StorageClass:  hostpath-provisioner
Status:        Bound
Volume:        pv0002
Labels:        alerts.k8s.io/KubePersistentVolumeFillingUp=disabled
               app=containerized-data-importer
               app.kubernetes.io/component=storage
               app.kubernetes.io/managed-by=cdi-controller
Annotations:   cdi.kubevirt.io/storage.bind.immediate.requested: true
               cdi.kubevirt.io/storage.condition.bound: true
               cdi.kubevirt.io/storage.condition.bound.message:
               cdi.kubevirt.io/storage.condition.bound.reason:
               cdi.kubevirt.io/storage.condition.running: true
               cdi.kubevirt.io/storage.condition.running.message:
               cdi.kubevirt.io/storage.condition.running.reason: Pod is running
               cdi.kubevirt.io/storage.contentType: kubevirt
               cdi.kubevirt.io/storage.deleteAfterCompletion: true
               cdi.kubevirt.io/storage.pod.phase: Running
               cdi.kubevirt.io/storage.pod.ready: true
               cdi.kubevirt.io/storage.pod.restarts: 0
               cdi.kubevirt.io/storage.preallocation.requested: false
               cdi.kubevirt.io/storage.upload.target:
               cdi.kubevirt.io/storage.uploadPodName: cdi-upload-vsrx-boot
               pv.kubernetes.io/bind-completed: yes
               pv.kubernetes.io/bound-by-controller: yes
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:      25Gi
Access Modes:  RWO
VolumeMode:    Filesystem <<<<<<<<<<===========================<<
Used By:       cdi-upload-vsrx-boot
Events:        <none>
root@pf-cloud$
root@pf-cloud$ kubectl describe pvc/vsrx-boot-scratch
Name:          vsrx-boot-scratch
Namespace:     default
StorageClass:  hostpath-provisioner
Status:        Bound
Volume:        pv0001
Labels:        app=containerized-data-importer
               app.kubernetes.io/component=storage
               app.kubernetes.io/managed-by=cdi-controller
Annotations:   pv.kubernetes.io/bind-completed: yes
               pv.kubernetes.io/bound-by-controller: yes
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:      25Gi
Access Modes:  RWO
VolumeMode:    Filesystem. <<<<<<<<<<===========================<<
Used By:       cdi-upload-vsrx-boot
Events:        <none>
root@pf-cloud$
root@pf-cloud$ kubectl get storageclass
NAME                             PROVISIONER                          RECLAIMPOLICY   VOLUMEBINDINGMODE      ALLOWVOLUMEEXPANSION   AGE
hostpath-provisioner (default)   kubernetes.io/hostpath-provisioner   Delete          WaitForFirstConsumer   false                  24h
root@pf-cloud$

root@pf-cloud$ kubectl describe storageclass/hostpath-provisioner
Name:            hostpath-provisioner
IsDefaultClass:  Yes
Annotations:     kubectl.kubernetes.io/last-applied-configuration={"apiVersion":"storage.k8s.io/v1","kind":"StorageClass","metadata":{"annotations":{"storageclass.kubernetes.io/is-default-class":"true"},"name":"hostpath-provisioner"},"provisioner":"kubernetes.io/hostpath-provisioner","reclaimPolicy":"Delete","volumeBindingMode":"WaitForFirstConsumer"}
,storageclass.kubernetes.io/is-default-class=true
Provisioner:           kubernetes.io/hostpath-provisioner
Parameters:            <none>
AllowVolumeExpansion:  <unset>
MountOptions:          <none>
ReclaimPolicy:         Delete
VolumeBindingMode:     WaitForFirstConsumer
Events:                <none>

**Comment by awels** [comment]
Okay so you are using the hostpath provisioner, but it looks like you are using the legacy version. What is your output when you run `kubectl get hostpathprovisioner -n hostpath-provisioner -o yaml`

**Comment by awels** [comment]
It should look something like this
```yaml
apiVersion: hostpathprovisioner.kubevirt.io/v1beta1
kind: HostPathProvisioner
metadata:
  name: hostpath-provisioner
spec:
  imagePullPolicy: Always
  storagePools:
    - name: "local"
      path: "/var/hpvolumes"
  workload:
    nodeSelector:
      kubernetes.io/os: linux
```
and not like this
```yaml
apiVersion: hostpathprovisioner.kubevirt.io/v1beta1
kind: HostPathProvisioner
metadata:
  name: hostpath-provisioner
spec:
  imagePullPolicy: IfNotPresent
  pathConfig:
    path: "/var/hpvolumes"
    useNamingPrefix: false
  workload:
    nodeSelector:
      kubernetes.io/os: linux
```
The second one deploys the legacy version of the provisioner, and you need to do some selinux relabeling to make it work, the new version is the csi version and that one works without relabeling.

**Comment by MaisenbacherD** [comment]
I bumped into the same issue a few weeks ago. Turns out there is a fix for it:
https://github.com/kubevirt/containerized-data-importer/pull/2676

Just install CDI v1.54.3 which includes that patch. :)

**Comment by manaspal2** [comment]
> I bumped into the same issue a few weeks ago. Turns out there is a fix for it: #2676
> 
> Just install CDI v1.54.3 which includes that patch. :)

Actually i followed manual approach. I logged in to the worker node. 
Changed the permission of hostPath as per the pv. But thanks for sharing the information.
